lib/view_model/callbacks.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+require 'renum'
+
+# Callback hooks for viewmodel traversal contexts
+module ViewModel::Callbacks
+  extend ActiveSupport::Concern
+
+  # Define the possible callback hooks and their required parameters.
+  enum :Hook do
+    BeforeVisit(:context)
+    AfterVisit(:context)
+
+    # The before deserialize hook is called when the viewmodel is visited during
+    # deserialization. At this point we don't know whether deserialization will
+    # be making any changes.
+    BeforeDeserialize(:deserialize_context)
+
+    # The on change hook is called if deserialization has visited the model and
+    # intends to make changes. Keyword argument `changes` is a
+    # ViewModel::Changes describing the intention. Callbacks on this hook may
+    # not themselves make any changes to the model. ViewModels backed by a
+    # transactional model such as AR may not have been saved, to allow the hook
+    # to inspect initial values.
+    OnChange(:deserialize_context, :changes)
+
+    AfterDeserialize(:deserialize_context)
+
+    attr_reader :context_name, :required_params
+
+    def init(context_name, *other_params)
+      @context_name = context_name
+      @required_params = [context_name, *other_params]
+    end
+
+    def dsl_name
+      name.underscore
+    end
+  end
+
+  # Placeholder for callbacks to invoked for all view types
+  ALWAYS = "__always"
```

Comment:
Discussed out of band: at least in principle view names are constrained in the same way as Ruby type names, so this cannot collide as is. We might want to add explicit constraints for view names, since I think we don't have them yet.